mypy/undefined_vars.py

```diff
@@ -0,0 +1,183 @@
+    WhileStmt,
+)
+from mypy.traverser import TraverserVisitor
+
+
+class DefinedVariables(NamedTuple):
+    """DefinedVariables contains information about variable definition at the end of a branching statement.
+    `if` and `match` are examples of branching statements.
+
+    `may_be_defined` contains variables that were defined in only some branches.
+    `must_be_defined` contains variables that were defined in all branches.
+    """
+
+    may_be_defined: set[str]
+    must_be_defined: set[str]
+
+
+class BranchingTracker:
+    def __init__(self) -> None:
+        self.defined_by_branch: list[DefinedVariables] = [
+            DefinedVariables(may_be_defined=set(), must_be_defined=set())
+        ]
+
+    def next_branch(self) -> None:
+        self.defined_by_branch.append(
+            DefinedVariables(may_be_defined=set(), must_be_defined=set())
+        )
+
+    def record_definition(self, name: str) -> None:
+        assert len(self.defined_by_branch) > 0
+        self.defined_by_branch[-1].must_be_defined.add(name)
+        self.defined_by_branch[-1].may_be_defined.discard(name)
+
+    def record_nested_branch(self, vars: DefinedVariables) -> None:
+        assert len(self.defined_by_branch) > 0
+        current_branch = self.defined_by_branch[-1]
+        current_branch.must_be_defined.update(vars.must_be_defined)
+        current_branch.may_be_defined.update(vars.may_be_defined)
+        current_branch.may_be_defined.difference_update(current_branch.must_be_defined)
+
+    def is_possibly_undefined(self, name: str) -> bool:
+        assert len(self.defined_by_branch) > 0
+        return name in self.defined_by_branch[-1].may_be_defined
+
+    def done(self) -> DefinedVariables:
+        assert len(self.defined_by_branch) > 0
+        if len(self.defined_by_branch) == 1:
+            # If there's only one branch, then we just return current.
+            # Note that this case is a different case when an empty branch is omitted (e.g. `if` without `else`).
+            return self.defined_by_branch[0]
+
+        # must_be_defined is a union of must_be_defined of all branches.
+        must_be_defined = set(self.defined_by_branch[0].must_be_defined)
+        for branch_vars in self.defined_by_branch[1:]:
+            must_be_defined.intersection_update(branch_vars.must_be_defined)
+        # may_be_defined are all variables that are not must be defined.
+        all_vars = set()
+        for branch_vars in self.defined_by_branch:
+            all_vars.update(branch_vars.may_be_defined)
+            all_vars.update(branch_vars.must_be_defined)
+        may_be_defined = all_vars.difference(must_be_defined)
+        return DefinedVariables(may_be_defined=may_be_defined, must_be_defined=must_be_defined)
+
+
+class DefinedVariableTracker:
+    """DefinedVariableTracker manages the state and scope for the UndefinedVariablesVisitor."""
+
+    def __init__(self) -> None:
+        # todo(stas): we should initialize this with some variables.
```

Comment:
I'd like to evaluate imports in some way in order to avoid false positives. Currently, this won't work:

```
from foo import x

if bool():
  x = 10
y = x  # This will incorrectly generate an error.
```

Is there a good way of doing this? Anywhere I should start looking?

Reply:
Can you treat `from foo import x` similarly to the assignment `x = ...` (the value of `x` doesn't make a difference to the analysis)?

Reply:
That would work some of the time. However, this doesn't work for `*` imports (i.e. `from foo import *`). After semantic analyzer, mypy should know which symbols are being imported, so it *should be* possible to get this info somehow.